File: thirtyinch/src/test/java/net/grandcentrix/thirtyinch/TiLifecycleObserverTest.java
/*
 * Copyright (C) 2017 grandcentrix GmbH
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.grandcentrix.thirtyinch;

import static org.assertj.core.api.Assertions.*;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyBoolean;
import static org.mockito.Mockito.*;

import java.util.ArrayList;
import java.util.List;
import org.junit.*;
import org.mockito.*;
import org.mockito.invocation.*;
import org.mockito.stubbing.*;

public class TiLifecycleObserverTest {

    private TiMockPresenter mPresenter;

    private TiView mView;

    @Before
    public void setUp() throws Exception {
        TiLog.setLogger(new TiLog.Logger() {
            @Override
            public void log(final int level, final String tag, final String msg) {
                // prevent RuntimeException: android.util.Log not mocked
            }
        });
        mView = mock(TiView.class);
        mPresenter = new TiMockPresenter();
    }

    @After
    public void tearDown() throws Exception {
        mPresenter = null;
        mView = null;
    }

    @Test
    public void testCalledAttachedInCorrectOrder() throws Exception {
        mPresenter.create();

        // Given 2 observers
        final TiLifecycleObserver observer1 = mock(TiLifecycleObserver.class);
        mPresenter.addLifecycleObserver(observer1);
        final TiLifecycleObserver observer2 = mock(TiLifecycleObserver.class);
        mPresenter.addLifecycleObserver(observer2);

        // When a view attaches
        mPresenter.attachView(mock(TiView.class));

        // Then the last added observer gets called last
        final InOrder inOrder = inOrder(observer1, observer2);
        inOrder.verify(observer1).onChange(TiPresenter.State.VIEW_ATTACHED, false);
        inOrder.verify(observer2).onChange(TiPresenter.State.VIEW_ATTACHED, false);

        inOrder.verify(observer1).onChange(TiPresenter.State.VIEW_ATTACHED, true);
        inOrder.verify(observer2).onChange(TiPresenter.State.VIEW_ATTACHED, true);
        inOrder.verifyNoMoreInteractions();
    }

    @Test
    public void testCalledCreateInCorrectOrder() throws Exception {

        // Given 2 observers
        final TiLifecycleObserver observer1 = mock(TiLifecycleObserver.class);
        mPresenter.addLifecycleObserver(observer1);
        final TiLifecycleObserver observer2 = mock(TiLifecycleObserver.class);
        mPresenter.addLifecycleObserver(observer2);

        // When the presenter gets created and reached view detached state
        mPresenter.create();

        // Then the last added observer gets called first because it's a destructive event
        final InOrder inOrder = inOrder(observer1, observer2);
        inOrder.verify(observer2).onChange(TiPresenter.State.VIEW_DETACHED, false);
        inOrder.verify(observer1).onChange(TiPresenter.State.VIEW_DETACHED, false);

        inOrder.verify(observer2).onChange(TiPresenter.State.VIEW_DETACHED, true);
        inOrder.verify(observer1).onChange(TiPresenter.State.VIEW_DETACHED, true);
        inOrder.verifyNoMoreInteractions();
    }

    @Test
    public void testCalledDestroyInCorrectOrder() throws Exception {

        // Given a presenter with 2 added observers
        mPresenter.create();
        final TiLifecycleObserver observer1 = mock(TiLifecycleObserver.class);
        mPresenter.addLifecycleObserver(observer1);
        final TiLifecycleObserver observer2 = mock(TiLifecycleObserver.class);
        mPresenter.addLifecycleObserver(observer2);

        // When the presenter gets destroyed
        mPresenter.destroy();

        // Then the last added observer gets called first
        final InOrder inOrder = inOrder(observer1, observer2);
        inOrder.verify(observer2).onChange(TiPresenter.State.DESTROYED, false);
        inOrder.verify(observer1).onChange(TiPresenter.State.DESTROYED, false);

        inOrder.verify(observer2).onChange(TiPresenter.State.DESTROYED, true);
        inOrder.verify(observer1).onChange(TiPresenter.State.DESTROYED, true);
        inOrder.verifyNoMoreInteractions();
    }

    @Test
    public void testCalledDetachedInCorrectOrder() throws Exception {
        mPresenter.create();
        mPresenter.attachView(mock(TiView.class));

        // Given 2 observers
        final TiLifecycleObserver observer1 = mock(TiLifecycleObserver.class);
        mPresenter.addLifecycleObserver(observer1);
        final TiLifecycleObserver observer2 = mock(TiLifecycleObserver.class);
        mPresenter.addLifecycleObserver(observer2);

        // When the view detached
        mPresenter.detachView();

        // Then the last added observer gets called first
        final InOrder inOrder = inOrder(observer1, observer2);
        inOrder.verify(observer2).onChange(TiPresenter.State.VIEW_DETACHED, false);
        inOrder.verify(observer1).onChange(TiPresenter.State.VIEW_DETACHED, false);

        inOrder.verify(observer2).onChange(TiPresenter.State.VIEW_DETACHED, true);
        inOrder.verify(observer1).onChange(TiPresenter.State.VIEW_DETACHED, true);
    }

    @Test
    public void testCreate() throws Exception {
        final List<Object[]> states = new ArrayList<>();
        mPresenter.addLifecycleObserver(new TiLifecycleObserver() {
            @Override
            public void onChange(final TiPresenter.State state,
                    final boolean hasLifecycleMethodBeenCalled) {
                states.add(new Object[]{state, hasLifecycleMethodBeenCalled});
            }
        });

        mPresenter.create();

        final Object[] beforeLast = states.get(states.size() - 2);
        assertThat(beforeLast[0]).isEqualTo(TiPresenter.State.VIEW_DETACHED);
        assertThat(beforeLast[1]).isEqualTo(false);

        final Object[] last = states.get(states.size() - 1);
        assertThat(last[0]).isEqualTo(TiPresenter.State.VIEW_DETACHED);
        assertThat(last[1]).isEqualTo(true);
    }

    @Test
    public void testDestroy() throws Exception {
        final List<Object[]> states = new ArrayList<>();
        mPresenter.addLifecycleObserver(new TiLifecycleObserver() {
            @Override
            public void onChange(final TiPresenter.State state,
                    final boolean hasLifecycleMethodBeenCalled) {
                states.add(new Object[]{state, hasLifecycleMethodBeenCalled});
            }
        });

        assertThat(mPresenter.mLifecycleObservers).hasSize(1);

        mPresenter.create();
        mPresenter.attachView(mView);
        mPresenter.detachView();
        mPresenter.destroy();

        final Object[] beforeLast = states.get(states.size() - 2);
        assertThat(beforeLast[0]).isEqualTo(TiPresenter.State.DESTROYED);
        assertThat(beforeLast[1]).isEqualTo(false);

        final Object[] last = states.get(states.size() - 1);
        assertThat(last[0]).isEqualTo(TiPresenter.State.DESTROYED);
        assertThat(last[1]).isEqualTo(true);

        assertThat(mPresenter.mLifecycleObservers).isEmpty();
    }

    @Test
    public void testRemoveObserver() throws Exception {
        final List<Object[]> states = new ArrayList<>();
        final Removable removable = mPresenter.addLifecycleObserver(new TiLifecycleObserver() {
            @Override
            public void onChange(final TiPresenter.State state,
                    final boolean hasLifecycleMethodBeenCalled) {
                states.add(new Object[]{state, hasLifecycleMethodBeenCalled});
            }
        });

        mPresenter.create();

        final Object[] beforeLast = states.get(states.size() - 2);
        assertThat(beforeLast[0]).isEqualTo(TiPresenter.State.VIEW_DETACHED);
        assertThat(beforeLast[1]).isEqualTo(false);

        final Object[] last = states.get(states.size() - 1);
        assertThat(last[0]).isEqualTo(TiPresenter.State.VIEW_DETACHED);
        assertThat(last[1]).isEqualTo(true);

        removable.remove();

        mPresenter.attachView(mView);

        final Object[] beforeLast2 = states.get(states.size() - 2);
        assertThat(beforeLast2[0]).isNotEqualTo(TiPresenter.State.VIEW_ATTACHED);

        final Object[] last2 = states.get(states.size() - 1);
        assertThat(last2[0]).isNotEqualTo(TiPresenter.State.VIEW_ATTACHED);
    }

    @Test
    public void testRemoveObserverTwice() throws Exception {
        final List<Object[]> states = new ArrayList<>();
        final TiLifecycleObserver observer = new TiLifecycleObserver() {
            @Override
            public void onChange(final TiPresenter.State state,
                    final boolean hasLifecycleMethodBeenCalled) {
                states.add(new Object[]{state, hasLifecycleMethodBeenCalled});
            }
        };

        final Removable removable = mPresenter.addLifecycleObserver(observer);
        assertThat(mPresenter.mLifecycleObservers).hasSize(1);
        removable.remove();
        assertThat(mPresenter.mLifecycleObservers).isEmpty();

        final Removable removable2 = mPresenter.addLifecycleObserver(observer);
        // remove should only remove once
        removable.remove();
        assertThat(mPresenter.mLifecycleObservers).hasSize(1);

        removable2.remove();
        assertThat(mPresenter.mLifecycleObservers).isEmpty();
    }

    @Test
    public void testRemoveOtherObserver() throws Exception {
        mPresenter.create();

        // add observers only for attach event
        final TiLifecycleObserver observer1 = mock(TiLifecycleObserver.class);
        mPresenter.addLifecycleObserver(observer1);
        final TiLifecycleObserver observer2 = mock(TiLifecycleObserver.class);
        final Removable removable = mPresenter.addLifecycleObserver(observer2);

        // when observer1 receives the first event it unregisters observer2
        doAnswer(new Answer() {
            @Override
            public Object answer(final InvocationOnMock invocation) throws Throwable {
                removable.remove();
                return null;
            }
        }).when(observer1).onChange(any(TiPresenter.State.class), anyBoolean());

        mPresenter.attachView(mock(TiView.class));

        final InOrder inOrder = inOrder(observer1, observer2);

        //observer 1 receives pre onAttachView event
        inOrder.verify(observer1).onChange(TiPresenter.State.VIEW_ATTACHED, false);

        // observer2 receives the pre event even when observer1 removed observer2 before observer2 received the pre event
        inOrder.verify(observer2).onChange(TiPresenter.State.VIEW_ATTACHED, false);

        // observer 1 receives post onAttachView event
        inOrder.verify(observer1).onChange(TiPresenter.State.VIEW_ATTACHED, true);

        // observer2 never receives the post event, is unregistered at that time
        verifyNoMoreInteractions(observer2);
    }

    @Test
    public void testSleep() throws Exception {
        final List<Object[]> states = new ArrayList<>();
        mPresenter.addLifecycleObserver(new TiLifecycleObserver() {
            @Override
            public void onChange(final TiPresenter.State state,
                    final boolean hasLifecycleMethodBeenCalled) {
                states.add(new Object[]{state, hasLifecycleMethodBeenCalled, mPresenter.getView()});
            }
        });

        mPresenter.create();
        mPresenter.attachView(mView);
        mPresenter.detachView();

        final Object[] beforeLast = states.get(states.size() - 2);
        assertThat(beforeLast[0]).isEqualTo(TiPresenter.State.VIEW_DETACHED);
        assertThat(beforeLast[1]).isEqualTo(false);
        assertThat(beforeLast[2]).isNotNull();

        final Object[] last = states.get(states.size() - 1);
        assertThat(last[0]).isEqualTo(TiPresenter.State.VIEW_DETACHED);
        assertThat(last[1]).isEqualTo(true);
        assertThat(last[2]).isNotNull();
    }

    @Test
    public void testWakeup() throws Exception {
        final List<Object[]> states = new ArrayList<>();
        mPresenter.addLifecycleObserver(new TiLifecycleObserver() {
            @Override
            public void onChange(final TiPresenter.State state,
                    final boolean hasLifecycleMethodBeenCalled) {
                states.add(new Object[]{state, hasLifecycleMethodBeenCalled, mPresenter.getView()});
            }
        });

        mPresenter.create();
        mPresenter.attachView(mView);

        final Object[] beforeLast = states.get(states.size() - 2);
        assertThat(beforeLast[0]).isEqualTo(TiPresenter.State.VIEW_ATTACHED);
        assertThat(beforeLast[1]).isEqualTo(false);
        assertThat(beforeLast[2]).isNotNull();

        final Object[] last = states.get(states.size() - 1);
        assertThat(last[0]).isEqualTo(TiPresenter.State.VIEW_ATTACHED);
        assertThat(last[1]).isEqualTo(true);
        assertThat(last[2]).isNotNull();
    }
}

File: thirtyinch/src/test/java/net/grandcentrix/thirtyinch/TiLifecycleObserverTest.kt
/*
 * Copyright (C) 2017 grandcentrix GmbH
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.grandcentrix.thirtyinch

import io.mockk.MockKAnnotations
import io.mockk.impl.annotations.MockK
import io.mockk.mockk
import io.mockk.verifySequence
import org.assertj.core.api.Assertions.*
import org.junit.*
import org.mockito.*
import org.mockito.Mockito.*

class TiLifecycleObserverTest {

    private lateinit var presenter: TiMockPresenter

    @MockK
    lateinit var view: TiView

    @Before
    fun setUp() {
        MockKAnnotations.init(this)
        TiLog.setLogger { _, _, _ ->
            // prevent RuntimeException: android.util.Log not mocked
        }
        presenter = TiMockPresenter()
    }

    @Test
    fun `test called attached in correct order`() {
        presenter.create()

        // Given 2 observers
        val observer1 = mockk<TiLifecycleObserver>(relaxUnitFun = true)
        presenter.addLifecycleObserver(observer1)
        val observer2 = mockk<TiLifecycleObserver>(relaxUnitFun = true)
        presenter.addLifecycleObserver(observer2)

        // When a view attaches
        presenter.attachView(mockk())

        // Then the last added observer gets called last
        verifySequence {
            observer1.onChange(TiPresenter.State.VIEW_ATTACHED, false)
            observer2.onChange(TiPresenter.State.VIEW_ATTACHED, false)

            observer1.onChange(TiPresenter.State.VIEW_ATTACHED, true)
            observer2.onChange(TiPresenter.State.VIEW_ATTACHED, true)
        }
    }

    @Test
    fun `test called create in correct order`() {
        // Given 2 observers
        val observer1 = mockk<TiLifecycleObserver>(relaxUnitFun = true)
        presenter.addLifecycleObserver(observer1)
        val observer2 = mockk<TiLifecycleObserver>(relaxUnitFun = true)
        presenter.addLifecycleObserver(observer2)

        // When the presenter gets created and reached view detached state
        presenter.create()

        // Then the last added observer gets called first because it's a destructive event
        verifySequence {
            observer2.onChange(TiPresenter.State.VIEW_DETACHED, false)
            observer1.onChange(TiPresenter.State.VIEW_DETACHED, false)

            observer2.onChange(TiPresenter.State.VIEW_DETACHED, true)
            observer1.onChange(TiPresenter.State.VIEW_DETACHED, true)
        }
    }

    @Test
    fun `test called destroy in correct order`() {
        // Given a presenter with 2 added observers
        presenter.create()
        val observer1 = mockk<TiLifecycleObserver>(relaxUnitFun = true)
        presenter.addLifecycleObserver(observer1)
        val observer2 = mockk<TiLifecycleObserver>(relaxUnitFun = true)
        presenter.addLifecycleObserver(observer2)

        // When the presenter gets destroyed
        presenter.destroy()

        // Then the last added observer gets called first
        verifySequence {
            observer2.onChange(TiPresenter.State.DESTROYED, false)
            observer1.onChange(TiPresenter.State.DESTROYED, false)

            observer2.onChange(TiPresenter.State.DESTROYED, true)
            observer1.onChange(TiPresenter.State.DESTROYED, true)
        }
    }

    @Test
    fun `test called detached in correct order`() {
        presenter.create()
        presenter.attachView(mockk())

        // Given 2 observers
        val observer1 = mockk<TiLifecycleObserver>(relaxUnitFun = true)
        presenter.addLifecycleObserver(observer1)
        val observer2 = mockk<TiLifecycleObserver>(relaxUnitFun = true)
        presenter.addLifecycleObserver(observer2)

        // When the view detached
        presenter.detachView()

        // Then the last added observer gets called first
        verifySequence {
            observer2.onChange(TiPresenter.State.VIEW_DETACHED, false)
            observer1.onChange(TiPresenter.State.VIEW_DETACHED, false)

            observer2.onChange(TiPresenter.State.VIEW_DETACHED, true)
            observer1.onChange(TiPresenter.State.VIEW_DETACHED, true)
        }
    }

    @Test
    fun `test create`() {
        val states = mutableListOf<Pair<TiPresenter.State, Boolean>>()
        presenter.addLifecycleObserver { state, hasLifecycleMethodBeenCalled ->
            states.add(Pair(state, hasLifecycleMethodBeenCalled))
        }

        presenter.create()

        val beforeLast = states[states.size - 2]
        assertThat(beforeLast.first).isEqualTo(TiPresenter.State.VIEW_DETACHED)
        assertThat(beforeLast.second).isEqualTo(false)

        val last = states.last()
        assertThat(last.first).isEqualTo(TiPresenter.State.VIEW_DETACHED)
        assertThat(last.second).isEqualTo(true)
    }

    @Test
    fun `test destroy`() {
        val states = mutableListOf<Pair<TiPresenter.State, Boolean>>()
        presenter.addLifecycleObserver { state, hasLifecycleMethodBeenCalled ->
            states.add(Pair(state, hasLifecycleMethodBeenCalled))
        }

        assertThat(presenter.mLifecycleObservers).hasSize(1)

        presenter.create()
        presenter.attachView(view)
        presenter.detachView()
        presenter.destroy()

        val beforeLast = states[states.size - 2]
        assertThat(beforeLast.first).isEqualTo(TiPresenter.State.DESTROYED)
        assertThat(beforeLast.second).isEqualTo(false)

        val last = states.last()
        assertThat(last.first).isEqualTo(TiPresenter.State.DESTROYED)
        assertThat(last.second).isEqualTo(true)

        assertThat(presenter.mLifecycleObservers).isEmpty()
    }

    @Test
    fun `test remove observer`() {
        val states = mutableListOf<Pair<TiPresenter.State, Boolean>>()
        val removable = presenter.addLifecycleObserver { state, hasLifecycleMethodBeenCalled ->
            states.add(Pair(state, hasLifecycleMethodBeenCalled))
        }

        presenter.create()

        val beforeLast = states[states.size - 2]
        assertThat(beforeLast.first).isEqualTo(TiPresenter.State.VIEW_DETACHED)
        assertThat(beforeLast.second).isEqualTo(false)

        val last = states.last()
        assertThat(last.first).isEqualTo(TiPresenter.State.VIEW_DETACHED)
        assertThat(last.second).isEqualTo(true)

        removable.remove()

        presenter.attachView(view)

        val beforeLast2 = states[states.size - 2]
        assertThat(beforeLast2.first).isNotEqualTo(TiPresenter.State.VIEW_ATTACHED)

        val last2 = states.last()
        assertThat(last2.first).isNotEqualTo(TiPresenter.State.VIEW_ATTACHED)
    }

    @Test
    fun `test remove observer twice`() {
        val states = mutableListOf<Pair<TiPresenter.State, Boolean>>()
        val observer = TiLifecycleObserver { state, hasLifecycleMethodBeenCalled ->
            states.add(Pair(state, hasLifecycleMethodBeenCalled))
        }

        val removable = presenter.addLifecycleObserver(observer)
        assertThat(presenter.mLifecycleObservers).hasSize(1)
        removable.remove()
        assertThat(presenter.mLifecycleObservers).isEmpty()

        val removable2 = presenter.addLifecycleObserver(observer)
        // remove should only remove once
        removable.remove()
        assertThat(presenter.mLifecycleObservers).hasSize(1)

        removable2.remove()
        assertThat(presenter.mLifecycleObservers).isEmpty()
    }

    @Test
    fun `test remove other observer`() {
        presenter.create()

        // add observers only for attach event
        val observer1 = mock(TiLifecycleObserver::class.java)
        presenter.addLifecycleObserver(observer1)
        val observer2 = mock(TiLifecycleObserver::class.java)
        val removable = presenter.addLifecycleObserver(observer2)

        // when observer1 receives the first event it unregisters observer2
        doAnswer {
            removable.remove()
            null
        }.`when`(observer1).onChange(
                ArgumentMatchers.any(TiPresenter.State::class.java),
                ArgumentMatchers.anyBoolean()
        )

        presenter.attachView(mock(TiView::class.java))

        val inOrder = inOrder(observer1, observer2)

        //observer 1 receives pre onAttachView event
        inOrder.verify(observer1).onChange(TiPresenter.State.VIEW_ATTACHED, false)

        // observer2 receives the pre event even when observer1 removed observer2 before observer2 received the pre event
        inOrder.verify(observer2).onChange(TiPresenter.State.VIEW_ATTACHED, false)

        // observer 1 receives post onAttachView event
        inOrder.verify(observer1).onChange(TiPresenter.State.VIEW_ATTACHED, true)

        // observer2 never receives the post event, is unregistered at that time
        verifyNoMoreInteractions(observer2)
    }

    @Test
    fun `test sleep`() {
        val states = mutableListOf<Triple<TiPresenter.State, Boolean, TiView?>>()
        presenter.addLifecycleObserver { state, hasLifecycleMethodBeenCalled ->
            states.add(Triple(state, hasLifecycleMethodBeenCalled, presenter.view))
        }

        presenter.create()
        presenter.attachView(view)
        presenter.detachView()

        val beforeLast = states[states.size - 2]
        assertThat(beforeLast.first).isEqualTo(TiPresenter.State.VIEW_DETACHED)
        assertThat(beforeLast.second).isEqualTo(false)
        assertThat(beforeLast.third).isNotNull()

        val last = states.last()
        assertThat(last.first).isEqualTo(TiPresenter.State.VIEW_DETACHED)
        assertThat(last.second).isEqualTo(true)
        assertThat(last.third).isNotNull()
    }

    @Test
    fun `test wake up`() {
        val states = mutableListOf<Triple<TiPresenter.State, Boolean, TiView?>>()
        presenter.addLifecycleObserver { state, hasLifecycleMethodBeenCalled ->
            states.add(Triple(state, hasLifecycleMethodBeenCalled, presenter.view))
        }

        presenter.create()
        presenter.attachView(view)

        val beforeLast = states[states.size - 2]
        assertThat(beforeLast.first).isEqualTo(TiPresenter.State.VIEW_ATTACHED)
        assertThat(beforeLast.second).isEqualTo(false)
        assertThat(beforeLast.third).isNotNull()

        val last = states.last()
        assertThat(last.first).isEqualTo(TiPresenter.State.VIEW_ATTACHED)
        assertThat(last.second).isEqualTo(true)
        assertThat(last.third).isNotNull()
    }
}
